File: packages/typescriptlib/src/test/typescript/serviceFrameworkTestUtils.ts
import {BaseService, Service} from "@src/typescript/service";
import {ServiceConfiguration} from "@src/typescript/serviceConfiguration";
import {ServiceInstanceProvider} from "@src/typescript/serviceInstanceProvider";
import {ServiceLifecycleController} from "@src/typescript/serviceLifecycleController";
import {mock, MockProxy} from "vitest-mock-extended";

class MockServiceInstanceProvider implements ServiceInstanceProvider<MockProxy<Service>>
{
    createdTimes: number = 0;

    constructor(public mockService: MockProxy<Service>) {}

    createServiceInstance(): MockProxy<Service>
    {
        this.createdTimes++;
        return this.mockService;
    }
}

class MockService extends BaseService
{
};

class MockServiceLifecycleController<T extends Service> implements ServiceLifecycleController<T>
{
    constructor()
    {
        return mock<ServiceLifecycleController<T>>();
    }

    init(serviceInstanceProvider: ServiceInstanceProvider<T>, config: ServiceConfiguration): void
    {
        throw new Error("Method not implemented.");
    }
    shutdown(): void
    {
        throw new Error("Method not implemented.");
    }
    getService(): T
    {
        throw new Error("Method not implemented.");
    }
};

export {MockServiceLifecycleController, MockServiceInstanceProvider, MockService};

File: packages/pythonlib/src/test/python/service_manager/service_framework_test_utils.py
"""
    Utility classes and functions used during testing
"""

from service_instance_provider import ServiceInstanceProvider
from service_lifecycle_controller import ServiceLifecycleController

from service import BaseService, CConfiguration, TService


class MockServiceInstanceProvider(ServiceInstanceProvider[TService]):
    """
    Mock Service Instance Provider
    """

    def __init__(self, mock_service: TService):
        self.created_times = 0
        self.mock_service = mock_service

    def create_service_instance(self) -> TService:
        self.created_times += 1
        return self.mock_service


class MockService(BaseService):
    """
    Mock Service
    """

    pass


class MockServiceLifecycleController(
    ServiceLifecycleController[TService, CConfiguration]
):
    """
    Mock Service Lifecycle Manager
    """

    def __init__(self):
        # Placeholder for mocking
        pass

    def init(
        self,
        service_instance_provider: ServiceInstanceProvider[TService],
        config: CConfiguration,
    ) -> None:
        pass

    def shutdown(self) -> None:
        pass

    def get_service(self) -> TService:
        pass
